### PR TITLE
[FIX] mail: removes event listener on unmount

### PR DESCRIPTION
**Pull request by ThanhDodeurOdoo**
Before this commit, an event listener was still active after the removal
of the attachment viewer, which caused the the onLoad function to be
called on an removed component if the component was unmounted before
the image was loaded.

part of task-2363522

### CODE REVIEW

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 1]
```diff
@@ -60,6 +60,7 @@ class AttachmentViewer extends Component {
          */
         this._renderedAttachment = undefined;
         this._onClickGlobal = this._onClickGlobal.bind(this);
+        this._boundOnLoadImage = this._onLoadImage.bind(this);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:63]: Don't call the method "bound", it's fine to "replace" the original method, as is done with `_onClickGlobal` just above

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 2]
```diff
@@ -79,6 +80,7 @@ class AttachmentViewer extends Component {
 
     willUnmount() {
         document.removeEventListener('click', this._onClickGlobal);
+        this._imageRef.el && this._imageRef.el.removeEventListener('load', this._boundOnLoadImage);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:83]: Don't use `&&` followed by instruction, it's a bit confusing. Just make a proper `if`.

Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:83]: Also question: considering the listener is also added in `patched`, shouldn't it be also removed in `willPatch`?

**Review by seb-odoo — CHANGES_REQUESTED** [review]
See comments, also not possible to write a test?

**Comment by ThanhDodeurOdoo** [comment]
@seb-odoo I updated the branch, regarding a test I'm not sure how I would test it but since it's a programming mistake I don't think it should have a test, the same way we don't test the proper removal of listeners elsewhere.

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 3]
```diff
@@ -69,6 +70,12 @@ class AttachmentViewer extends Component {
         document.addEventListener('click', this._onClickGlobal);
     }
 
+    willPatch() {
+        if (this._imageRef.el) {
+            this._imageRef.el.removeEventListener('load', this._onLoadImage);
+        }
+    }
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:77]: This one should have a similar condition to the one in `_handleImageLoad`

`this._renderedAttachment !== this.attachmentViewer.attachment`
Because otherwise will might remove the listener but never add it back.

Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:77]: Alternative is to always remove and then re-add the listener, so we don't need to worry about such conditions. Not sure why we didn't in the first place, maybe it looked like an optimization?

cc @alexkuhn

Comment (alexkuhn) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:77]: > Alternative is to always remove and then re-add the listener, so we don't need to worry about such conditions. Not sure why we didn't in the first place, maybe it looked like an optimization?
> 
> cc @alexkuhn

Fine by me as fix. `_handleImageLoad` is clumsy code-reuse, but that's something to be cleaned in master.

Comment (ThanhDodeurOdoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:77]: @seb-odoo j'ai mis a jour la branche, j'ai aussi rajouté un test et j'ai vérifié qu'il ne passait pas sans le fix.

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 4]
```diff
@@ -157,11 +167,10 @@ class AttachmentViewer extends Component {
      */
     _handleImageLoad() {
         if (
-            this.attachmentViewer.attachment.fileType === 'image' &&
-            this._renderedAttachment !== this.attachmentViewer.attachment
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:161]: maybe revert condition change

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 5]
```diff
@@ -69,6 +70,12 @@ class AttachmentViewer extends Component {
         document.addEventListener('click', this._onClickGlobal);
     }
 
+    willPatch() {
+        if (this._imageRef.el) {
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:74]: and then maybe here, remove listener only if isImageLoading is false

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 6]
```diff
@@ -157,11 +167,10 @@ class AttachmentViewer extends Component {
      */
     _handleImageLoad() {
         if (
-            this.attachmentViewer.attachment.fileType === 'image' &&
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:160]: if filetype is not image, isImageLoading should be reset to false

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js (hunk 5 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:74]: or if filetype is not image anymore

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:161]: Or another alternative would be if we can detect if the image is already loaded or not, and add listener only if image is not loaded (and then we can still remove the other condition)

**Review by seb-odoo — APPROVED** [review]
I guess this is good now, @alexkuhn do you want final check?

**Comment by seb-odoo** [comment]
Actually I still notice a problem when I try. When I go from a loaded image to a yet to load image, the spinner does not show, and I actually still see the previous loaded image.

**Comment by ThanhDodeurOdoo** [comment]
> Actually I still notice a problem when I try. When I go from a loaded image to a yet to load image, the spinner does not show, and I actually still see the previous loaded image.

I see, it's odd indeed, I push a fix for that, it's the 14.0 behaviour.

**Comment by ThanhDodeurOdoo** [comment]
@alexkuhn

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 7]
```diff
@@ -30,17 +31,13 @@ class AttachmentViewer extends Component {
                 attachmentViewer: attachmentViewer ? attachmentViewer.__state : undefined,
             };
         });
+        this._getRefs = useRefs();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:34]: Add comment somewhere (probably here) why useRefs is needed instead of single ref (issue about t-key and t-ref)
And then add also comment why we need t-key in the first place (to force a new element because changing src doesn't work)

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 8]
```diff
@@ -156,12 +153,13 @@ class AttachmentViewer extends Component {
      * @private
      */
     _handleImageLoad() {
+        const refs = this._getRefs();
+        const image = refs[`image_${this.attachmentViewer.attachment.id}`];
         if (
             this.attachmentViewer.attachment.fileType === 'image' &&
-            this._renderedAttachment !== this.attachmentViewer.attachment
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:161]: I think the `_renderedAttachment` variable is useless now and can be removed

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml [hunk 9]
```diff
@@ -39,7 +39,7 @@
                                 <i class="fa fa-3x fa-circle-o-notch fa-fw fa-spin" role="img" title="Loading"/>
                             </div>
                         </t>
-                        <img class="o_AttachmentViewer_view o_AttachmentViewer_viewImage" t-on-click="_onClickImage" t-on-mousedown="_onMousedownImage" t-on-wheel="_onWheelImage" t-att-src="attachmentViewer.attachment.defaultSource" t-att-style="imageStyle" draggable="false" alt="Viewer" t-ref="image"/>
+                        <img class="o_AttachmentViewer_view o_AttachmentViewer_viewImage" t-on-click="_onClickImage" t-on-mousedown="_onMousedownImage" t-on-wheel="_onWheelImage" t-on-load="_onLoadImage" t-att-src="attachmentViewer.attachment.defaultSource" t-att-style="imageStyle" draggable="false" alt="Viewer" t-key="'image_' + attachmentViewer.attachment.id" t-ref="image_{{attachmentViewer.attachment.id}}"/>
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml:42]: It's a bit strange to use `+` to build the key but `{{ }}` to build the ref. Can't we use the same syntax for both?
Note: If you use the syntax of ref, should have space inside { }

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 10]
```diff
@@ -31,6 +31,10 @@ class AttachmentViewer extends Component {
                 attachmentViewer: attachmentViewer ? attachmentViewer.__state : undefined,
             };
         });
+        /**
+         * Used to ensure that the ref is always up to date, which seems to be needed if the element
+         * has a t-key, which was added to force the rendering of a new element when the src of the image changes.
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:36]: add the reason of the src change (-> show loading instead of previous image while it is loading)

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml (hunk 9 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml]: still mix of syntax here

Comment (ThanhDodeurOdoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml]: `t-ref` and `t-key` don't work the same way, `t-key` expects a string, `t-ref` expects a formatted expression

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 11]
```diff
@@ -156,12 +151,13 @@ class AttachmentViewer extends Component {
      * @private
      */
     _handleImageLoad() {
+        const refs = this._getRefs();
+        const image = refs[`image_${this.attachmentViewer.attachment.id}`];
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:155]: I think we can have a race condition here: if `attachmentViewer` or `attachment` is deleted just before the image is loaded, we might go into the `_handleImageLoad` before the component is destroyed (but models are already deleted).
It would be safer to guard those variables with preemptive `return`.

File: addons/mail/static/src/components/attachment/attachment_tests.js [hunk 12]
```diff
@@ -595,6 +595,34 @@ QUnit.test('close attachment viewer', async function (assert) {
...
+        document.querySelector('.o_AttachmentViewer_headerItemButtonClose').click()
+    );
+    image.dispatchEvent(new Event('load', { bubbles: true }));
+    assert.containsNone(
+        document.body,
+        '.o_Dialog',
+        "attachment viewer should be closed after clicking on close button"
+    );
+});
+
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment/attachment_tests.js:625]: It's short, sure, but the name doesn't give a spec. The purpose of test is rather vague.
Maybe this test crashed without fix, but this doesn't mean the test is good.

Can you tell exactly what feature didn't work without this fix? If it's the closing of dialog not working, then maybe the test itself is good and it's just the name is off. In that case, suggestion:

- Name: _"should close attachment viewer of image before image is loaded"_ (just a few additional words, and it makes purpose of test very explicit)
- Assert no attachment viewer rather than dialog. The dialog is just container

File: addons/mail/static/src/components/attachment/attachment_tests.js [hunk 13]
```diff
@@ -595,6 +595,34 @@ QUnit.test('close attachment viewer', async function (assert) {
...
+        detailsMode: 'hover',
+        isDownloadable: false,
+        isEditable: false,
+    });
+    await afterNextRender(() => document.querySelector('.o_Attachment_image').click());
+    const image = document.querySelector('.o_AttachmentViewer_viewImage');
+    await afterNextRender(() =>
+        document.querySelector('.o_AttachmentViewer_headerItemButtonClose').click()
+    );
+    image.dispatchEvent(new Event('load', { bubbles: true }));
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment/attachment_tests.js:618]: Add comment on what is simulated here.

Comment (alexkuhn) [inline addons/mail/static/src/components/attachment/attachment_tests.js:618]: Don't use variable `image`, especially when obtained before `afterNextRender()` and used `afterNextRender()`.

File: addons/mail/static/src/components/attachment/attachment_tests.js [hunk 14]
```diff
@@ -595,6 +595,34 @@ QUnit.test('close attachment viewer', async function (assert) {
...
+    const attachment = this.env.models['mail.attachment'].create({
+        filename: "test.png",
+        id: 750,
+        mimetype: 'image/png',
+        name: "test.png",
+    });
+    await this.createAttachmentComponent(attachment, {
+        detailsMode: 'hover',
+        isDownloadable: false,
+        isEditable: false,
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment/attachment_tests.js:611]: These props looks irrelevant for test. Not sure if `detailsMode` is, but maybe it is?

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/attachment/attachment_tests.js (hunk 12 above)
Comment (ThanhDodeur) [reply addons/mail/static/src/components/attachment/attachment_tests.js:625]: Without the fix, there is a traceback on 'load' dispatch.

Comment (ThanhDodeurOdoo) [reply addons/mail/static/src/components/attachment/attachment_tests.js:625]: Without the fix, there is a traceback (attachmentVIewer is undefined) when the 'load' event is dispatched, I changed for 0 assertions and a more explicit title.

File: addons/mail/static/src/components/attachment/attachment_tests.js (hunk 13 above)
Comment (ThanhDodeurOdoo) [reply addons/mail/static/src/components/attachment/attachment_tests.js:618]: It is on purpose, I need a reference to image before it is removed from the DOM, I add a comment for that.

File: addons/mail/static/src/components/attachment/attachment_tests.js (hunk 12 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/attachment/attachment_tests.js]: Hm... Don't really like this test, but considering the complexity with workaround using refs, it's fine to have a test.

Would add `[technical]` prefix on test name (with brief comment, it's good for `[technical]` tests to provide some assumption on how the code was designed at the moment of writing of test).

Rather than `assert.expect(0)`, I would prefer `assert.expect(1)` and use of `try/catch`. Adding assertion(s) where needed makes it super clear what the test is actually testing. Considering the test is about not crashing, it should be just `assert.ok()` at end of `try` or `assert.notOk` at end of `catch`.

File: addons/mail/static/src/components/attachment/attachment_tests.js (hunk 13 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/attachment/attachment_tests.js]: Ok, this piece of information could be provided in heading comment of this technical test too.

File: addons/mail/static/src/components/attachment/attachment_tests.js [hunk 15]
```diff
@@ -630,6 +630,35 @@ QUnit.test('clicking on the delete attachment button multiple times should do th
...
+    });
+    await this.createAttachmentComponent(attachment);
+    await afterNextRender(() => document.querySelector('.o_Attachment_image').click());
+    // getting a reference to the image before closing the viewer, otherwise the image can't be retrieved from
+    // the DOM as it is removed with the viewer
+    const image = document.querySelector('.o_AttachmentViewer_viewImage');
+    await afterNextRender(() =>
+        document.querySelector('.o_AttachmentViewer_headerItemButtonClose').click()
+    );
+    // simulates the image 'load' event, triggered when the image is rendered
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment/attachment_tests.js:653]:
```js
// Simulate image becoming loaded.
```

**Review by alexkuhn — CHANGES_REQUESTED** [review]
Almost there!

**Review by alexkuhn — APPROVED** [review]
@robodoo r+
